// gluestack-ui/dist/config.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.config = void 0;
    const config = {
        repoUrl: 'https://github.com/gluestack/gluestack-ui.git',
        branchName: 'patch',
        tagName: '',
        writableComponentsPath: 'components',
        style: '',
        gluestackStyleRootPath: 'themed',
        nativeWindRootPath: 'nativewind',
        providerComponent: 'gluestack-ui-provider',
        componentsResourcePath: 'example/storybook-nativewind/src/core-components',
        gluestackDir: '.gluestack/cache/gluestack-ui',
        configFileName: 'config.ts',
        UIconfigPath: 'components/gluestack-ui-provider/config.ts',
        nativeWindDependencies: ['@gluestack-ui/nativewind-utils'],
        tailwindConfigRootPath: 'example/storybook-nativewind/tailwind.config.js',
        nativewindUtilPattern: '@gluestack-ui/nativewind-utils/',
        gluestackUIPattern: '@gluestack-ui/',
        packageVersions: {
            'react-native-svg': '13.4.0',
            nativewind: '4.0.1',
        },
    };
    exports.config = config;
});
